### DISCUSSION TITLE
Streamplot from numerical data

### DISCUSSION THREAD

**Discussion by navdeeprana** [JuliaPlots/Makie.jl#3157] [Q&A]
The [documentation](https://docs.makie.org/stable/examples/plotting_functions/streamplot/) says that we can create a streamplot for a function that returns  `dx,dy` for a given `x,y`. Say I have numerical values of the coordinates and the velocities, how can I create a streamplot? Essentially, I want a plotting recipe equivalent to [matplotlib streamplot](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.streamplot.html).

**Comment by navdeeprana** [comment]
@jkrumbiegel, anything on this?

**Reply by SimonDanisch** [reply]
Do you want https://docs.makie.org/stable/examples/plotting_functions/arrows/index.html#arrows?

**Comment by jkrumbiegel** [comment]
I think we currently only take a function because you need to be able to evaluate it continuously. But we could (or you could) use an interpolation over the matrix as the function, which should then result in what matplotlib does.

**Comment by navdeeprana** [comment]
Okay, I will give it a shot.

**Comment by navdeeprana** [comment]
Okay, this simple snippet works.

```julia
using CairoMakie, Interpolations
function splot(u, v)
    nx, ny = size(u)
    x, y = 1:nx, 1:ny
    intu, intv = linear_interpolation((x,y), u), linear_interpolation((x,y), v)
    f(x) = Point2f(intu(x...), intv(x...))
    return streamplot(f, x, y, colormap=:magma)
end

# Streamplot example function
struct FitzhughNagumo{T}
    ϵ::T
    s::T
    γ::T
    β::T
end
P = FitzhughNagumo(0.1, 0.0, 1.5, 0.8)

x = -1.5:0.1:1.5
nx = length(x)
u, v = zeros(nx, nx), zeros(nx, nx)

for (j, xj) in enumerate(x)
    for (i, xi) in enumerate(x)
        u[i, j], v[i, j] = (xi-xj-xi^3+P.s)/P.ϵ, P.γ*xi-xj + P.β
    end
end
fig, ax, pl = splot(u, v)
```
Output
![image](https://github.com/MakieOrg/Makie.jl/assets/17153984/50880d59-4e8d-491e-b0b4-38dd4502652d)

Can we extend the recipe to support plotting numerical data? If yes, I can improve upon it and open a pull request. Any suggestions on improvements are also welcome.

**Reply by henry2004y** [reply]
I also want this to be directly supported by the streamplot recipe! This is one of last few missing pieces when I migrate my whole plotting functions from Matplotlib/Matlab to Makie.

**Reply by NessLange** [reply]
@navdeeprana That is a really cool Code! Do you may reveal how you added the arrows inside your plot 
and how to extend your example to **vector input** instead of **matrices** ?

As I see it **u** and **v** are matrices, but what if (like in my case) you got **x, y, u, v** vectors (same length)

**arrows(x,y,u,v)** works fine, but I don't figure what to change in your code to accept that input.

**Reply by navdeeprana** [reply]
@NessLange, thanks. The arrows are added by the `streamplot` function that is called inside `splot`. Check it's documentation for more details.

In splot, I assume that the input is matrix. But for 1D data, you can try if the following works
```julia
function splot(x, y, u, v)
    intu, intv = linear_interpolation((x,y), u), linear_interpolation((x,y), v)
    f(x) = Point2f(intu(x...), intv(x...))
    return streamplot(f, x, y, colormap=:magma)
end
```
I am on the phone so can't check it at the moment.

**Reply by NessLange** [reply]
Thanks for the quick reply. I have checked the nice `streamplot` documentation and ran the same exact code without having the arrows being drawn in my plot. 

Unfortunately I also get an error when running your modified code with vector input. 

Here are some data to tinker with

`n = 20;
x = 0:n; x = x'.*ones(n+1).*2*pi/n;
y = 0:n; y = ones(n+1)'.*y.*3*pi/n;
u =  sin.(x) .* cos.(y)
v = -cos.(x) .* sin.(y)
x = x[:]; y = y[:]; u = u[:]; v = v[:];
`

**Edit:** I do right know, in that case I can just leave the data as matrix and use your original code, but in the case of my real data it is not as easy to transform to a matrix.

**Reply by asinghvi17** [reply]
Maybe you want an interpolator like https://github.com/DanielVandH/NaturalNeighbours.jl?  Or something like this anyway.

You might also want to look at https://github.com/MakieOrg/TopoPlots.jl and see the interpolators used there!

**Reply by NessLange** [reply]
Awesome, these look like some really fruitful sources! I would  never have found that.

**Comment by navdeeprana** [comment]
About your arrows issue, I do not know, perhaps you have a modified Makie theme or some other custom setting, try updating Makie, running the example code in a clean repl/notebook.

About plotting 1D arrays, Interpolation over a D-tuple requires D dimensional data. For example this snippet
```julia
using CairoMakie, Interpolations
function splot(x, y, u, v)
    intu, intv = linear_interpolation((x,y), u), linear_interpolation((x,y), v)
    f(x) = Point2f(intu(x...), intv(x...))
    return streamplot(f, x, y, colormap=:magma)
end

x = 0:0.2:2π
y = 0:0.2:2π
u = @. sin(x) * cos(y')
v = @. cos(x) * sin(y')
splot(x, y, u, v)
```
yields the following plot
![image](https://github.com/MakieOrg/Makie.jl/assets/17153984/4b9f71b6-faaf-4922-a604-876df054edd5)



However `splot(x,y,vec(u),vec(v))` yields the following error
```
MethodError: no method matching linear_interpolation(::Tuple{StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}, StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}}, ::Vector{Float64})

Closest candidates are:
  linear_interpolation(::Tuple{Vararg{AbstractRange, N}}, ::AbstractArray{T, N}; extrapolation_bc) where {N, T}
   @ Interpolations ~/.julia/packages/Interpolations/91PhN/src/convenience-constructors.jl:22
  linear_interpolation(::Tuple{Vararg{AbstractVector, N}}, ::AbstractArray{T, N}; extrapolation_bc) where {N, T}
   @ Interpolations ~/.julia/packages/Interpolations/91PhN/src/convenience-constructors.jl:25
  linear_interpolation(::AbstractRange, ::AbstractVector; extrapolation_bc)
   @ Interpolations ~/.julia/packages/Interpolations/91PhN/src/convenience-constructors.jl:7
```
Note the function signature ` linear_interpolation(::Tuple{Vararg{AbstractRange, N}}, ::AbstractArray{T, N}; extrapolation_bc) where {N, T}`

So, I think you need to reshape your `u, v` arrays into matrices and then use `splot`. Note that matplotlib has similar requirements as mentioned [here](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.streamplot.html)

**Reply by NessLange** [reply]
alright, thank you :)
